apps/openmw/mwmechanics/spellcasting.cpp

```diff
@@ -44,26 +44,44 @@ namespace MWMechanics
         return schoolSkillMap[school];
     }
 
+    float calcEffectCost(ESM::ENAMstruct effect)
+    {
+        const ESM::MagicEffect* magicEffect = MWBase::Environment::get().getWorld()->getStore().get<ESM::MagicEffect>().find(effect.mEffectID);
+
+        int minMaxMagn = 2;
+        if (!(magicEffect->mData.mFlags & ESM::MagicEffect::NoMagnitude))
+            minMaxMagn = effect.mMagnMin + effect.mMagnMax;
+
+        int duration = 0;
+        if (!(magicEffect->mData.mFlags & ESM::MagicEffect::NoDuration))
+            duration = effect.mDuration;
+
+        if (!(magicEffect->mData.mFlags & ESM::MagicEffect::UncappedDamage))
+            duration = std::max(1, duration);
+
+        static const float fEffectCostMult = MWBase::Environment::get().getWorld()->getStore()
+            .get<ESM::GameSetting>().find("fEffectCostMult")->getFloat();
+
+        float cost = (minMaxMagn * (duration + 1) + std::max(1, effect.mArea)) * magicEffect->mData.mBaseCost * fEffectCostMult * 0.05;
```

Comment:
Best to avoid refactoring game mechanics code, renaming its variables, rearranging statements etc. It will become harder to verify that the code matches the [research page](https://wiki.openmw.org/index.php?title=Research:Trading_and_Services#Spell_merchant). 

Code duplication should always be avoided though so its OK that you made a new function.